metrics/metrics.go

```diff
@@ -0,0 +1,192 @@
+		)
+
+		certificateRequestRenewalTimeSeconds = prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "certificate_request_renewal_timestamp_seconds",
+				Help:      "The number of seconds before expiration time the certificate request should renew.",
+			},
+			[]string{"name", "namespace", "issuer_name", "issuer_kind", "issuer_group"},
+		)
+
+		certificateRequestReadyStatus = prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "certificate_request_ready_status",
+				Help:      "The ready status of the certificate request.",
+			},
+			[]string{"name", "namespace", "condition", "issuer_name", "issuer_kind", "issuer_group"},
+		)
+
+		driverIssueCallCount = prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "driver_issue_call_count",
+				Help:      "The number of issue() calls made by the driver.",
+			},
+			[]string{"node", "volume"},
+		)
+
+		driverIssueErrorCount = prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "driver_issue_error_count",
+				Help:      "The number of errors encountered during the driver issue() calls.",
+			},
+			[]string{"node", "volume"},
+		)
+
+		managedVolumeCount = prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "managed_volume_count",
+				Help:      "The number of volume managed by the csi driver.",
+			},
+			[]string{"node"},
+		)
+
+		managedCertificateCount = prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "managed_certificate_count",
+				Help:      "The number of certificates managed by the csi driver.",
+			},
+			[]string{"node"},
+		)
+	)
+
+	// Create Registry and register the recommended collectors
+	registry := prometheus.NewRegistry()
+	registry.MustRegister(
+		collectors.NewProcessCollector(collectors.ProcessCollectorOpts{}),
+		collectors.NewGoCollector(),
+	)
+	// Create server and register Prometheus metrics handler
+	m := &Metrics{
+		log:      logger.WithName("metrics"),
+		registry: registry,
+
+		certificateRequestExpiryTimeSeconds:  certificateRequestExpiryTimeSeconds,
+		certificateRequestRenewalTimeSeconds: certificateRequestRenewalTimeSeconds,
+		certificateRequestReadyStatus:        certificateRequestReadyStatus,
+		driverIssueCallCount:                 driverIssueCallCount,
+		driverIssueErrorCount:                driverIssueErrorCount,
+		managedVolumeCount:                   managedVolumeCount,
+		managedCertificateCount:              managedCertificateCount,
+	}
+
+	return m
+}
+
+// NewServer registers Prometheus metrics and returns a new Prometheus metrics HTTP server.
+func (m *Metrics) NewServer(ln net.Listener) *http.Server {
```

Comment:
This isn't called outside of test cases, and I guess that is by design as it is expected that the corresponding implementation should call `NewServer` on `metrics.Metrics` to register their Listener.

Could you possibly update the `example/` implementation in the root of this repository to demonstrate how to actually add the `/metrics` endpoint? I also wonder if the Managed should be extended to be able to auto-serve this endpoint in cases where a user doesn't _need_ to provide their own listener (but _does_ want metrics to be served).

Reply:
What if we renamed this to `Register(*prometheus.Registry)` rather than tying the `net.Listener` logic into the registration?

We can always have/find some kind of `csihelpers.Handle(*http.Server, *prometheus.Registry)` function elsewhere then, which needn't be opinionated about csi-lib.

Reply:
> Could you possibly update the example/ implementation in the root of this repository to demonstrate how to actually add the /metrics endpoint? 

Yes, it is there. And modified accordingly based on recent changes. 
https://github.com/7ing/csi-lib/blob/abf15631238fa809d10d9c206178c414d9495b4f/test/integration/metrics_test.go#L83-L95

Reply:
Thanks for the suggestions. I have changed to use a `DefaultHandler` instead, which could be served as a reference implementation for http handler.